### PR TITLE
Nonmappable sort keys

### PR DESCRIPTION
**Pull request by alissapajer**
Part of #3021

Provide support for nonmappable sort keys.

Add tests for `ExtractFreeMap`.

### CODE REVIEW

File: connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala [hunk 1]
```diff
@@ -71,24 +71,77 @@ final class ExtractFreeMap[T[_[_]]: BirecursiveT] private () extends QSUTTypes[T
...
+              }
+          }
         }
-      } map { nel => graph.overwriteAtRoot(QSSort(srcRoot, Nil, nel)) }
+      }
+
+      nonmappable match {
+        case Nil =>
+          keys traverse { case (k, sortDir) =>
+            MappableRegion.unaryOf(src.root, graph refocus k.root).strengthR(sortDir) match {
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala]: I shouldn't be running this again; instead we can use `access` from above. I'll make that change.

File: connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala [hunk 2]
```diff
@@ -71,24 +71,77 @@ final class ExtractFreeMap[T[_[_]]: BirecursiveT] private () extends QSUTTypes[T
...
+            QSUGraph(graph.root, newVerts)
+          }
+      }
     }
 
   def apply[F[_]: Monad: NameGenerator: PlannerErrorME](graph: QSUGraph)
       : F[QSUGraph] =
     graph.rewriteM[F](extract[F])
 
-  ////
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala]: I'll add that back, didn't mean to delete it.

File: it/src/main/resources/tests/sortKeyFlatten.test [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+     "backends": {
+         "couchbase": "ignoreFieldOrder",
+         "marklogic_json": "ignoreFieldOrder",
+         "mimir": "ignoreFieldOrder"
+     },
+
+    "data": "zips.data",
+
+    "query": "select city, state, loc from zips order by loc[*] desc",
```
Comment (alissapajer) [inline it/src/main/resources/tests/sortKeyFlatten.test]: All the integration tests were erroring locally, so maybe there is a typo in this test, it was taking too long to iterate through fixing this. This query works in the REPL though. We'll see what Travis says.

File: connector/src/test/scala/quasar/qscript/qsu/ExtractFreeMapSpec.scala [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                AutoJoin2(
+                  Read(`orders`),
+                  Transpose(Map(Read(`orders`), innerFM), Retain.Values, Rotation.ShiftMap),
+                  innerAutojoinCondition),
+                Transpose(Map(Read(`orders`), outerFM), Retain.Identities, Rotation.ShiftArray),
+                outerAutojoinCondition),
+              Nil,
+              NEL(
+                (fm1, SortDir.Ascending), ICons((fm2, SortDir.Descending), ICons((fm3, SortDir.Descending), ICons((fm4, SortDir.Ascending), INil()))))),
+            valueAccess)) =>  // FIXME don't explicitly test for generated names
```
Comment (alissapajer) [inline connector/src/test/scala/quasar/qscript/qsu/ExtractFreeMapSpec.scala]: We just need to test that each of the two places below where `"__fromTree3"` and `"__fromTree6"` show up, they are the same value.

File: it/src/main/resources/tests/sortKeyFlatten.test (hunk 3 above)
Comment (alissapajer) [reply it/src/main/resources/tests/sortKeyFlatten.test]: Oh cool the test passed! :)

File: connector/src/test/scala/quasar/qscript/qsu/ExtractFreeMapSpec.scala [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+          (innerFM must_= key2) and
```
Comment (djspiewak) [inline connector/src/test/scala/quasar/qscript/qsu/ExtractFreeMapSpec.scala]: You don't need the `and`s here (or the other places where you've added them).  The whole body of the partial function inside a `beLike` is treated just like a normal expectation, so sequential conjunction of assertions may be encoded just by having them on subsequent lines.

File: it/src/main/resources/tests/sortKeyFlatten.test (hunk 3 above)
Comment (djspiewak) [reply it/src/main/resources/tests/sortKeyFlatten.test]: Cool!  But related to that, there are at least two tests I found that have actual SQL^2 parse errors.

File: connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala [hunk 6]
```diff
@@ -71,24 +71,77 @@ final class ExtractFreeMap[T[_[_]]: BirecursiveT] private () extends QSUTTypes[T
...
+        case head :: tail =>
+          for {
+            interRoot <- freshName[F]
+            joined <- autojoinVerts(head, tail)
+          } yield {
+            val (joinRoot, verts) = joined
+
+            val order: NEL[(FreeMap, SortDir)] =
+              access.map(_.leftMap {
+                case -\/(fm) => fm >> func.ProjectKey(func.Hole, StrLit("sort_source"))
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/ExtractFreeMap.scala]: Not a necessary change, but just wanted to bring`func.ProjectKeyS` to your attention.

**Comment by wemrysi** [comment]
@alissapajer Ping me when you're finished with the changes you mentioned and I'll merge!

**Comment by alissapajer** [comment]
@wemrysi ready for another look!

**Pull request merged**